src/main/grammar/SourcePawn.flex

```diff
@@ -5,125 +5,191 @@ import static org.idea_sp.psi.SourcePawnTypes.*;
 
 %%
 
-%{
-  public _SourcePawnLexer() {
-    this((java.io.Reader)null);
-  }
-%}
-
 %public
 %class _SourcePawnLexer
 %implements FlexLexer
 %function advance
 %type IElementType
 %unicode
+%line
+%column
+
+%{
+  public _SourcePawnLexer() {
+    this((java.io.Reader)null);
+  }
+
+  int yyline;
+  int yycolumn;
+  int elementStart;
+  int tmp;
+  IElementType commentType;
+%}
 
 EOL="\r"|"\n"|"\r\n"
 LINE_WS=[\ \t\f]
 WHITE_SPACE=({LINE_WS}|{EOL})+
 
 SPACE=[\ \n\r\t\f]
 LINE_COMMENT="//"[^\r\n]*
-BLOCK_COMMENT="/"\*(.|\n)*?\*"/"
-PREPROCESSOR_COMMENT=#(assert|define|else|elseif|endif|endinput|error|file|if|include|line|pragma|section|tryinclude|undef)[^\r\n]*
-INTEGER_LITERAL=[-+]?[0-9][_\d]*
-FLOAT_LITERAL=[-+]?[0-9][_\d]*\.[0-9][_\d]*(e[-+]?[0-9]+)?
-HEX_LITERAL=[-+]?0x[a-fA-F0-9_]+
-BINARY_LITERAL=[-+]?0b[01_]+
+BLOCK_COMMENT_DOC_START="/**"
+BLOCK_COMMENT_START="/*"
+BLOCK_COMMENT_END="*/"
+PREPROCESSOR_COMMENT_START=#(assert|define|else|elseif|endif|endinput|error|file|if|include|line|pragma|section|tryinclude|undef)
+INTEGER_LITERAL=(\b[-+])?[0-9][_\d]*
+FLOAT_LITERAL=(\b[-+])?[0-9][_\d]*\.[0-9][_\d]*(e[-+]?[0-9]+)?
+HEX_LITERAL=(\b[-+])?0x[a-fA-F0-9_]+
+BINARY_LITERAL=(\b[-+])?0b[01_]+
 STRING_LITERAL=\"(\\.|[^\"])*\"
-CHARACTER_LITERAL='(\\.|[^\"])'
+CHARACTER_LITERAL='(\\.|[^'])'
 SYMBOL=([@_a-zA-Z][@_a-zA-Z0-9]+|[a-zA-Z][@_a-zA-Z0-9]*)
 
+%state COMMENT, P_COMMENT
+
 %%
+<COMMENT> {
+  {BLOCK_COMMENT_END}          { yybegin(YYINITIAL);
+                                 zzStartRead = elementStart;
+                                 return commentType; }
+  [^]                          { }
+}
+
+<P_COMMENT> {
+  {BLOCK_COMMENT_DOC_START}    { tmp = elementStart;
+                                 elementStart = zzCurrentPos;
+                                 commentType = DOC_COMMENT;
+                                 yybegin(COMMENT);
+                                 zzMarkedPos -= yylength();
+                                 zzStartRead = tmp;
+                                 return PREPROCESSOR_COMMENT; }
```

Comment:
all this business of screwing with lexer positionings is pretty darned gross. there is likely a much simpler and prettier way of doing this. At the very least, this hair could be extracted out to a method.